types/wlr_drm_lease_v1.c

```diff
@@ -0,0 +1,645 @@
+
+static void drm_lease_device_v1_handle_create_lease_request(
+		struct wl_client *client, struct wl_resource *resource, uint32_t id) {
+	struct wlr_drm_lease_device_v1 *device =
+		drm_lease_device_v1_from_resource(resource);
+
+	struct wlr_drm_lease_request_v1 *req =
+		calloc(1, sizeof(struct wlr_drm_lease_request_v1));
+	if (!req) {
+		wlr_log(WLR_ERROR, "Failed to allocate wlr_drm_lease_request_v1");
+		wl_resource_post_no_memory(resource);
+		return;
+	}
+
+	wlr_log(WLR_DEBUG, "Created request %p", req);
+
+	uint32_t version = wl_resource_get_version(resource);
+	struct wl_resource *request_resource = wl_resource_create(client,
+		&wp_drm_lease_request_v1_interface, version, id);
+	if (!request_resource) {
+		wlr_log(WLR_ERROR, "Failed to allocate wl_resource");
+		wl_resource_post_no_memory(resource);
+		free(req);
+		return;
+	}
+
+	req->device = device;
+	req->resource = request_resource;
+	req->connectors = NULL;
+	req->n_connectors = 0;
+
+	wl_resource_set_implementation(request_resource, &lease_request_impl,
+		req, drm_lease_request_v1_handle_resource_destroy);
+
+	wl_list_insert(&device->requests, wl_resource_get_link(request_resource));
+}
+
+static struct wp_drm_lease_device_v1_interface lease_device_impl = {
+	.stop = drm_lease_device_v1_handle_stop,
+	.create_lease_request = drm_lease_device_v1_handle_create_lease_request,
+};
+
+static void drm_connector_v1_handle_resource_destroy(
+		struct wl_resource *resource) {
+	struct wlr_drm_lease_connector_v1 *connector =
+		drm_lease_connector_v1_from_resource(resource);
+	drm_lease_connector_v1_destroy(connector);
+}
+
+static void drm_connector_v1_handle_destroy(
+		struct wl_client *client, struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static struct wp_drm_lease_connector_v1_interface lease_connector_impl = {
+	.destroy = drm_connector_v1_handle_destroy,
+};
+
+static void drm_lease_connector_v1_send_to_client(
+		struct wlr_drm_lease_connector_v1 *connector,
+		struct wl_resource *resource) {
+	if (connector->active_lease) {
+		return;
+	}
+
+	struct wl_client *client = wl_resource_get_client(resource);
+
+	uint32_t version = wl_resource_get_version(resource);
+	struct wl_resource *connector_resource = wl_resource_create(client,
+			&wp_drm_lease_connector_v1_interface, version, 0);
+	if (!connector_resource) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	wl_resource_set_implementation(connector_resource, &lease_connector_impl,
+			connector, drm_connector_v1_handle_resource_destroy);
+	wp_drm_lease_device_v1_send_connector(resource, connector_resource);
+
+	struct wlr_output *output = connector->output;
+	wp_drm_lease_connector_v1_send_name(connector_resource, output->name);
+
+	wp_drm_lease_connector_v1_send_description(connector_resource,
+			output->description);
+
+	struct wlr_drm_connector *drm_connector =
+			get_drm_connector_from_output(output);
+	wp_drm_lease_connector_v1_send_connector_id(connector_resource,
+			drm_connector->id);
```

Comment:
No need to use the private `get_drm_connector_from_output` function here, instead we can call `wlr_drm_connector_get_id`.